specs/language/basic.tex

```diff
@@ -1,5 +1,97 @@
 \Ch{Basic Concepts}{Basic}
 
+\Sec{Types}{Basic.types}
+
+\p The \textit{object representation} of an object of type \texttt{T} is the
+sequence of \textit{N} bytes taken up by the object of type \texttt{T}, where
+\textit{N} equals \texttt{sizeof(T)}\footnote{\texttt{sizeof(T)}, returns the
+size of an object of type \texttt{T} in bytes as stored in device memory.
+The size may be different in other memory spaces and no operator exists to query
+it.}. The \textit{object representation} of an object may be different based on
+the \textit{memory space} it is stored in (\ref{Intro.Memory.Spaces}).
```

Comment:
The wording in this footnote is confusing. I think it's saying
> "sizeof" returns the size of the struct as-if it's stored in device memory, and determining the size if it's stored in another memory space is not possible

Reply:
Yea, I think one of the problems with our definition of `sizeof` is that we don't allow you to specify which address space you want the size for. The value returned from `sizeof` is probably valid for most memory spaces, but it isn't valid for constant. For example:

```hlsl
// In a cbuffer this structure is 3 rows because of the
// packing rules, which makes it 48 bytes instead of 16.
struct FourDWords {
  float2 V;
  float A[2];
};
_Static_assert(sizeof(FourDWords) == sizeof(uint) * 4, "What you expect");
```

Also `sizeof` fails to compile if the type is a "special" type that only has a thread representation. For example:

```hlsl
struct SpecialType {
  uint U;
  RWBuffer<uint> B;
};
// error: invalid application of 'sizeof' to non-numeric type 'SpecialType'
_Static_assert(sizeof(SpecialType) == sizeof(uint), "Fails to compile");
```